### PR TITLE
Let the ajax requests (specially the ones saving data) finish before

### LINKED ISSUE #24303: LDAP password field needs to lose focus to save password and be able to connect.
### Steps to reproduce:

Having encryption enabled do:
1. Enable ldap app.
2. Set up ldap server as usual, add all requested data but DN and click on Detect Base DN.
### Expected behaviour:

LDAP server is configured and its users appear.
### Actual behaviour

Errors while setting it up.
### Server configuration

**Operating system**:
Ubuntu 14.04

**Web server:**
Apache

**Database:**
MySQL

**PHP version:**
5.5.9

**ownCloud version:** (see ownCloud admin page)
{"installed":true,"maintenance":false,"version":"9.0.2.0","versionstring":"9.0.2 RC1","edition":"Enterprise"}

Observed also in 8.2.4rc1

**Updated from an older ownCloud or fresh install:**
Fresh

**The content of config/config.php:**

```

```

**Are you using external storage, if yes which one:** local/smb/sftp/...
No

**Are you using encryption:** 
No.

**Logs**

```
{"reqId":"dhWqAyr3eAE+7Z2wzE7b","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Host given!","level":2,"time":"2016-04-27T08:23:13+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"dhWqAyr3eAE+7Z2wzE7b","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Port given!","level":2,"time":"2016-04-27T08:23:13+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"dhWqAyr3eAE+7Z2wzE7b","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Login Filter given!","level":2,"time":"2016-04-27T08:23:13+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"dhWqAyr3eAE+7Z2wzE7b","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): Not a single Base DN given.","level":2,"time":"2016-04-27T08:23:13+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"dhWqAyr3eAE+7Z2wzE7b","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): login filter does not contain %uid place holder.","level":2,"time":"2016-04-27T08:23:13+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"dhWqAyr3eAE+7Z2wzE7b","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Host given!","level":2,"time":"2016-04-27T08:23:13+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"dhWqAyr3eAE+7Z2wzE7b","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Port given!","level":2,"time":"2016-04-27T08:23:13+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"dhWqAyr3eAE+7Z2wzE7b","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Login Filter given!","level":2,"time":"2016-04-27T08:23:13+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"dhWqAyr3eAE+7Z2wzE7b","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): Not a single Base DN given.","level":2,"time":"2016-04-27T08:23:13+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"dhWqAyr3eAE+7Z2wzE7b","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): login filter does not contain %uid place holder.","level":2,"time":"2016-04-27T08:23:13+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"+FYkcC7cAwq0wqa0JdIY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Port given!","level":2,"time":"2016-04-27T08:23:16+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"+FYkcC7cAwq0wqa0JdIY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Login Filter given!","level":2,"time":"2016-04-27T08:23:16+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"+FYkcC7cAwq0wqa0JdIY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): Not a single Base DN given.","level":2,"time":"2016-04-27T08:23:16+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"+FYkcC7cAwq0wqa0JdIY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): login filter does not contain %uid place holder.","level":2,"time":"2016-04-27T08:23:16+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"+FYkcC7cAwq0wqa0JdIY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Port given!","level":2,"time":"2016-04-27T08:23:16+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"+FYkcC7cAwq0wqa0JdIY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Login Filter given!","level":2,"time":"2016-04-27T08:23:16+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"+FYkcC7cAwq0wqa0JdIY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): Not a single Base DN given.","level":2,"time":"2016-04-27T08:23:16+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"+FYkcC7cAwq0wqa0JdIY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): login filter does not contain %uid place holder.","level":2,"time":"2016-04-27T08:23:16+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"TpVhIIsEcNsM9ahgmeMY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Login Filter given!","level":2,"time":"2016-04-27T08:23:20+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"TpVhIIsEcNsM9ahgmeMY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): Not a single Base DN given.","level":2,"time":"2016-04-27T08:23:20+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"TpVhIIsEcNsM9ahgmeMY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): login filter does not contain %uid place holder.","level":2,"time":"2016-04-27T08:23:20+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"TpVhIIsEcNsM9ahgmeMY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Login Filter given!","level":2,"time":"2016-04-27T08:23:20+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"TpVhIIsEcNsM9ahgmeMY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): Not a single Base DN given.","level":2,"time":"2016-04-27T08:23:20+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"TpVhIIsEcNsM9ahgmeMY","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): login filter does not contain %uid place holder.","level":2,"time":"2016-04-27T08:23:20+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"G9L2bfvfQgLJ5EvWfDLE","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Login Filter given!","level":2,"time":"2016-04-27T08:23:28+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"G9L2bfvfQgLJ5EvWfDLE","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): Not a single Base DN given.","level":2,"time":"2016-04-27T08:23:28+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"G9L2bfvfQgLJ5EvWfDLE","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): login filter does not contain %uid place holder.","level":2,"time":"2016-04-27T08:23:28+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"G9L2bfvfQgLJ5EvWfDLE","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Login Filter given!","level":2,"time":"2016-04-27T08:23:28+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"G9L2bfvfQgLJ5EvWfDLE","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): Not a single Base DN given.","level":2,"time":"2016-04-27T08:23:28+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"G9L2bfvfQgLJ5EvWfDLE","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): login filter does not contain %uid place holder.","level":2,"time":"2016-04-27T08:23:28+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"uuXn\/dyQrUzHBNgm6ydu","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Login Filter given!","level":2,"time":"2016-04-27T08:23:32+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"uuXn\/dyQrUzHBNgm6ydu","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): either no password is given for theuser agent or a password is given, but not anLDAP agent.","level":2,"time":"2016-04-27T08:23:32+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"uuXn\/dyQrUzHBNgm6ydu","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): Not a single Base DN given.","level":2,"time":"2016-04-27T08:23:32+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"uuXn\/dyQrUzHBNgm6ydu","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): login filter does not contain %uid place holder.","level":2,"time":"2016-04-27T08:23:32+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"uuXn\/dyQrUzHBNgm6ydu","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Login Filter given!","level":2,"time":"2016-04-27T08:23:32+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"uuXn\/dyQrUzHBNgm6ydu","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): either no password is given for theuser agent or a password is given, but not anLDAP agent.","level":2,"time":"2016-04-27T08:23:32+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"uuXn\/dyQrUzHBNgm6ydu","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): Not a single Base DN given.","level":2,"time":"2016-04-27T08:23:32+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"uuXn\/dyQrUzHBNgm6ydu","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): login filter does not contain %uid place holder.","level":2,"time":"2016-04-27T08:23:32+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"tbayJKiGLtRl81ht8oav","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Login Filter given!","level":2,"time":"2016-04-27T08:23:33+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"tbayJKiGLtRl81ht8oav","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): either no password is given for theuser agent or a password is given, but not anLDAP agent.","level":2,"time":"2016-04-27T08:23:33+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"tbayJKiGLtRl81ht8oav","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): Not a single Base DN given.","level":2,"time":"2016-04-27T08:23:33+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"tbayJKiGLtRl81ht8oav","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): login filter does not contain %uid place holder.","level":2,"time":"2016-04-27T08:23:33+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"tbayJKiGLtRl81ht8oav","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): No LDAP Login Filter given!","level":2,"time":"2016-04-27T08:23:33+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"tbayJKiGLtRl81ht8oav","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): either no password is given for theuser agent or a password is given, but not anLDAP agent.","level":2,"time":"2016-04-27T08:23:33+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"tbayJKiGLtRl81ht8oav","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): Not a single Base DN given.","level":2,"time":"2016-04-27T08:23:33+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
{"reqId":"tbayJKiGLtRl81ht8oav","remoteAddr":"HIDDEN_IP","app":"user_ldap","message":"Configuration Error (prefix ): login filter does not contain %uid place holder.","level":2,"time":"2016-04-27T08:23:33+00:00","method":"POST","url":"\/index.php\/apps\/user_ldap\/ajax\/wizard.php","user":"admin"}
```
### Client configuration

**browser**
Chrome

### PR DESCRIPTION
**Pull request by jvillafanez**
getting data or doing some checks in the server

Currently only available for the LDAP "server" tab and the 3 buttons
there

Not the best solution, but I think this is the least intrusive one. We should only take into acount the save requests from the LDAP wizard instead of any ajax requests, but it doesn't seem easy to get to that point.

Fixes https://github.com/owncloud/core/issues/24303 (for master)

### CODE REVIEW

File: apps/user_ldap/js/wizard/wizardTabElementary.js [hunk 1]
```diff
@@ -66,9 +69,39 @@ OCA = OCA || {};
...
+				self.activeAjaxRequests = true;
+			});
+			$(document).on('ajaxStop.ajaxCounter', function(){
+				self.activeAjaxRequests = false;
+			});
+		},
+
+		runAfterAjaxRequests: function(func, args, thisContext) {
+			if (thisContext.activeAjaxRequests) {
+				var randomId = Math.random().toString().slice(2);
```
Comment (PVince81) [inline apps/user_ldap/js/wizard/wizardTabElementary.js]:
```js
var randomId = _.uniqueId('ajaxRequest');
```

File: apps/user_ldap/js/wizard/wizardTabElementary.js [hunk 2]
```diff
@@ -66,9 +69,39 @@ OCA = OCA || {};
...
+				// there are ajax requests running very likely one to save data, so wait until all
+				// requests finish
+				$(document).on('ajaxStop.' + randomId, function(){
+					// send the click event to the function
+					func.apply(thisContext, args);
```
Comment (PVince81) [inline apps/user_ldap/js/wizard/wizardTabElementary.js]: I don't really understand what you are doing here. Looks like some weird hackery.

File: apps/user_ldap/js/wizard/wizardTabElementary.js [hunk 3]
```diff
@@ -66,9 +69,39 @@ OCA = OCA || {};
...
+			});
+			$(document).on('ajaxStart.ajaxCounter', function(){
+				self.activeAjaxRequests = true;
+			});
+			$(document).on('ajaxStop.ajaxCounter', function(){
+				self.activeAjaxRequests = false;
+			});
+		},
+
+		runAfterAjaxRequests: function(func, args, thisContext) {
```
Comment (PVince81) [inline apps/user_ldap/js/wizard/wizardTabElementary.js]: This might work better by using Promises

**Comment by PVince81** [comment]
Also I don't really understand where this connects to real ajax requests.

Also note that ajax requests from the page are not limited to LDAP. There's the notifications ajax request, heartbeat, etc. So you are likely to get false positives.

:-1: so far, I think we need another approach.

**Comment by jvillafanez** [comment]
Unless the server is bombarded with a lot of requests, the solution should be enough. In case the heartbeat or any other call is being executed, this call will be queued and executed after. I don't think there are expensive call that will degrade the UX noticeable.

**Comment by PVince81** [comment]
Yeah maybe, but it would be better to try and use promises.
I thought the LDAP code had all the ajax calls in a central place where one might be able to hook in. :frowning:

**Comment by jvillafanez** [comment]
I'll have a look again on Monday before making any change

**Comment by jvillafanez** [comment]
I think we can close this in favor of https://github.com/owncloud/core/pull/25080

**Pull request closed without merging**